libevmasm/SemanticInformation.cpp

```diff
@@ -479,6 +479,8 @@ bool SemanticInformation::invalidInPureFunctions(Instruction _instruction)
 	case Instruction::EXTCODESIZE:
 	case Instruction::EXTCODECOPY:
 	case Instruction::EXTCODEHASH:
+	case Instruction::RETURNDATASIZE:
+	case Instruction::RETURNDATACOPY:
```

Comment:
I'm not sure what should or should not be pure. AFAIK, there is no proper definition for what is a pure function in EVM. So this is entirely arbitrary.

The only restriction I can think about is related to compile time constant evaluation. Which we don't properly have yet. If that is indeed the case, there should be more instructions here.

Reply:
Well, we don't even have full agreement if `pure` should refer to stuff that can be evaluated at compilation time or to stuff that does not touch state (#8153). So I think this out of scope of this PR.

This particular change to RETURNDATACOPY was explicitly requested by @chriseth:

> 0age just mentioned it the ethsecurity chat that returndatacopt is not marked as invalid in pure functions - can someone please create an issue about it and potentially even fix?

What I'm not clear about is whether this should really be considered a bug. I'd consider it a breaking change but from that comment it eeems to me like @chriseth would rather see it as a bugfix.

Reply:
Hmm.. why is this actually needed?

Reply:
Ahh.. I guess because `pure` is considered as being evaluable at compile time, right?

Reply:
Not exactly. Currently `pure` is closer to "not reading or writing", though there are discussions about changing/clarifying that. Reading returndata is a bit borderline IMO. I could see it being allowed.

Reply:
Yeah, seem to depend on how `pure` is exactly defined. I also understood it (now) as you said - not reading or writing from state (after reading the docs again). Isn't it correct to see it like that `RETURNDATACOPY` & `RETURNDATASIZE` is not reading or writing any state? 

Why should `RETURNDATACOPY` & `RETURNDATASIZE` not be `pure`? What is the big advantage to change that? Or isn't it like that `RETURNDATACOPY` & `RETURNDATASIZE`  can basically only be used, if a call was done before? `STATICCALL` is already considered as reading the state - it is already not allowed in pure functions. 

If `RETURNDATACOPY` & `RETURNDATASIZE` is used without a call (e.g. in pure functions) I see it like that `RETURNDATACOPY` will just not do anything. Where `RETURNDATASIZE` will always return `0`.

For all other cases, calls are needed, that are not allowed in `pure` functions.

Reply:
After having a short discussion with @chriseth  and @hrkrshnn, I finally understand the underlying issue. The problem is basically, that it is possible to do a call before calling a pure function that maybe using `RETURNDATACOPY` and `RETURNDATASIZE`. In this case the values depend on the previous calls.

Thats why it makes totally sense to disallow `RETURNDATACOPY` and `RETURNDATASIZE` in pure functions.